crates/loro-internal/src/handler.rs

```diff
@@ -4,115 +4,87 @@ use crate::{
     container::{
         idx::ContainerIdx,
         list::list_op::{DeleteSpan, DeleteSpanWithId, ListOp},
-        richtext::richtext_state::PosType,
+        richtext::{richtext_state::PosType, RichtextState, StyleOp, TextStyleInfoFlag},
         tree::tree_op::TreeOp,
     },
     delta::{DeltaItem, StyleMeta, TreeDiffItem, TreeExternalDiff},
     op::ListSlice,
-    state::{State, TreeParentId},
+    state::{ContainerState, State, TreeParentId},
     txn::EventHint,
     utils::{string_slice::StringSlice, utf16::count_utf16_len},
 };
+use append_only_bytes::BytesSlice;
 use enum_as_inner::EnumAsInner;
+use enum_dispatch::enum_dispatch;
 use fxhash::FxHashMap;
 use loro_common::{
-    ContainerID, ContainerType, InternalString, LoroError, LoroResult, LoroTreeError, LoroValue,
-    TreeID,
+    ContainerID, ContainerType, Counter, IdFull, InternalString, LoroError, LoroResult,
+    LoroTreeError, LoroValue, PeerID, TreeID, ID,
 };
 use serde::{Deserialize, Serialize};
 use std::{
     borrow::Cow,
+    fmt::Debug,
     ops::Deref,
-    sync::{Mutex, Weak},
+    sync::{Arc, Mutex, Weak},
 };
 
-#[derive(Debug, Clone, EnumAsInner, Deserialize, Serialize, PartialEq)]
-#[serde(untagged)]
-pub enum TextDelta {
-    Retain {
-        retain: usize,
-        attributes: Option<FxHashMap<String, LoroValue>>,
-    },
-    Insert {
-        insert: String,
-        attributes: Option<FxHashMap<String, LoroValue>>,
-    },
-    Delete {
-        delete: usize,
-    },
-}
-
-impl From<&DeltaItem<StringSlice, StyleMeta>> for TextDelta {
-    fn from(value: &DeltaItem<StringSlice, StyleMeta>) -> Self {
-        match value {
-            crate::delta::DeltaItem::Retain { retain, attributes } => TextDelta::Retain {
-                retain: *retain,
-                attributes: attributes.to_option_map(),
-            },
-            crate::delta::DeltaItem::Insert { insert, attributes } => TextDelta::Insert {
-                insert: insert.to_string(),
-                attributes: attributes.to_option_map(),
-            },
-            crate::delta::DeltaItem::Delete {
-                delete,
-                attributes: _,
-            } => TextDelta::Delete { delete: *delete },
-        }
-    }
-}
-
-pub trait HandlerTrait {
-    fn inner(&self) -> &BasicHandler;
-
-    fn get_value(&self) -> LoroValue {
-        self.inner()
-            .state
-            .upgrade()
-            .unwrap()
-            .lock()
-            .unwrap()
-            .get_value_by_idx(self.inner().container_idx)
-    }
+const INSERT_CONTAINER_VALUE_ARG_ERROR: &str =
+    "Cannot insert a LoroValue::Container directly. To create child container, use insert_container";
+
+#[enum_dispatch]
```

Comment:
Maybe it's unnecessary

Reply:
I think the error message will take less bytes in the binary if two strings are mostly the same with only a few differences. This can avoid that.

Reply:
haha. The line 35.

Reply:
Oh, done